File: test/py/web_app.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Redis 哨兵模式 Web 应用
使用 Flask 框架
"""

from flask import Flask, request, jsonify
import json
import time
from redis_config import get_redis_master, get_redis_slave, get_redis_info

app = Flask(__name__)

# 统一响应格式
def success_response(data=None):
    """成功响应"""
    return jsonify({
        'code': 200,
        'message': 'success',
        'data': data
    })

def error_response(code, message):
    """错误响应"""
    return jsonify({
        'code': code,
        'message': message
    }), code

@app.route('/api/v1/kv', methods=['POST'])
def set_key_value():
    """设置键值对"""
    try:
        data = request.get_json()
        key = data.get('key')
        value = data.get('value')
        ttl = data.get('ttl', 0)
        
        if not key or not value:
            return error_response(400, '键名和值不能为空')
        
        master = get_redis_master()
        
        if ttl > 0:
            master.setex(key, ttl, value)
        else:
            master.set(key, value)
        
        return success_response({
            'key': key,
            'value': value,
            'ttl': ttl
        })
        
    except Exception as e:
        return error_response(500, f'设置失败: {str(e)}')

@app.route('/api/v1/kv/<key>', methods=['GET'])
def get_key_value(key):
    """获取键值对"""
    try:
        slave = get_redis_slave()
        value = slave.get(key)
        
        if value is None:
            return error_response(404, '键不存在')
        
        return success_response({
            'key': key,
            'value': value
        })
        
    except Exception as e:
        return error_response(500, f'获取失败: {str(e)}')

@app.route('/api/v1/kv/<key>', methods=['DELETE'])
def delete_key(key):
    """删除键"""
    try:
        master = get_redis_master()
        count = master.delete(key)
        
        return success_response({
            'key': key,
            'count': count
        })
        
    except Exception as e:
        return error_response(500, f'删除失败: {str(e)}')

@app.route('/api/v1/kv/<key>/exists', methods=['GET'])
def exists_key(key):
    """检查键是否存在"""
    try:
        slave = get_redis_slave()
        exists = slave.exists(key)
        
        return success_response({
            'key': key,
            'exists': bool(exists)
        })
        
    except Exception as e:
        return error_response(500, f'检查失败: {str(e)}')

@app.route('/api/v1/users', methods=['POST'])
def set_user():
    """设置用户信息"""
    try:
        data = request.get_json()
        user_id = data.get('id')
        name = data.get('name')
        age = data.get('age')
        
        if not all([user_id, name, age]):
            return error_response(400, '用户ID、姓名和年龄不能为空')
        
        master = get_redis_master()
        user_key = f'user:{user_id}'
        
        user_data = {
            'id': user_id,
            'name': name,
            'age': age
        }
        
        master.hset(user_key, mapping=user_data)
        
        return success_response({
            'user': user_data
        })
        
    except Exception as e:
        return error_response(500, f'设置用户失败: {str(e)}')

@app.route('/api/v1/users/<int:user_id>', methods=['GET'])
def get_user(user_id):
    """获取用户信息"""
    try:
        slave = get_redis_slave()
        user_key = f'user:{user_id}'
        user_data = slave.hgetall(user_key)
        
        if not user_data:
            return error_response(404, '用户不存在')
        
        # 转换数据类型
        user_data['id'] = int(user_data['id'])
        user_data['age'] = int(user_data['age'])
        
        return success_response({
            'user': user_data
        })
        
    except Exception as e:
        return error_response(500, f'获取用户失败: {str(e)}')

@app.route('/api/v1/lists', methods=['POST'])
def add_list_item():
    """添加列表项"""
    try:
        data = request.get_json()
        list_key = data.get('list_key')
        items = data.get('items', [])
        
        if not list_key or not items:
            return error_response(400, '列表键名和项目不能为空')
        
        master = get_redis_master()
        count = master.lpush(list_key, *items)
        
        return success_response({
            'list_key': list_key,
            'count': count,
            'items': items
        })
        
    except Exception as e:
        return error_response(500, f'添加列表项失败: {str(e)}')

@app.route('/api/v1/lists/<list_key>', methods=['GET'])
def get_list(list_key):
    """获取列表"""
    try:
        slave = get_redis_slave()
        items = slave.lrange(list_key, 0, -1)
        
        return success_response({
            'list_key': list_key,
            'items': items,
            'count': len(items)
        })
        
    except Exception as e:
        return error_response(500, f'获取列表失败: {str(e)}')

@app.route('/api/v1/lists/<list_key>/pop', methods=['DELETE'])
def pop_list_item(list_key):
    """弹出列表项"""
    try:
        master = get_redis_master()
        item = master.rpop(list_key)
        
        if item is None:
            return error_response(404, '列表为空')
        
        return success_response({
            'list_key': list_key,
            'item': item
        })
        
    except Exception as e:
        return error_response(500, f'弹出列表项失败: {str(e)}')

@app.route('/api/v1/sets', methods=['POST'])
def add_set_member():
    """添加集合成员"""
    try:
        data = request.get_json()
        set_key = data.get('set_key')
        members = data.get('members', [])
        
        if not set_key or not members:
            return error_response(400, '集合键名和成员不能为空')
        
        master = get_redis_master()
        count = master.sadd(set_key, *members)
        
        return success_response({
            'set_key': set_key,
            'count': count,
            'members': members
        })
        
    except Exception as e:
        return error_response(500, f'添加集合成员失败: {str(e)}')

@app.route('/api/v1/sets/<set_key>', methods=['GET'])
def get_set_members(set_key):
    """获取集合成员"""
    try:
        slave = get_redis_slave()
        members = list(slave.smembers(set_key))
        
        return success_response({
            'set_key': set_key,
            'members': members,
            'count': len(members)
        })
        
    except Exception as e:
        return error_response(500, f'获取集合成员失败: {str(e)}')

@app.route('/api/v1/redis/status', methods=['GET'])
def get_redis_status():
    """获取 Redis 状态"""
    try:
        info = get_redis_info()
        
        if not info:
            return error_response(500, '获取 Redis 状态失败')
        
        return success_response({
            'master_addr': info['master'],
            'slave_addrs': info['slaves'],
            'sentinel_info': info['sentinels']
        })
        
    except Exception as e:
        return error_response(500, f'获取 Redis 状态失败: {str(e)}')

@app.route('/api/v1/health', methods=['GET'])
def health_check():
    """健康检查"""
    try:
        master = get_redis_master()
        slave = get_redis_slave()
        
        # 测试主节点连接
        master.ping()
        
        # 测试从节点连接
        slave.ping()
        
        return success_response({
            'status': 'healthy',
            'timestamp': int(time.time())
        })
        
    except Exception as e:
        return error_response(503, f'健康检查失败: {str(e)}')

@app.route('/')
def index():
    """首页"""
    return jsonify({
        'message': 'Redis 哨兵模式 Web 应用',
        'version': '1.0.0',
        'endpoints': [
            'POST /api/v1/kv - 设置键值对',
            'GET /api/v1/kv/:key - 获取键值对',
            'DELETE /api/v1/kv/:key - 删除键',
            'GET /api/v1/kv/:key/exists - 检查键是否存在',
            'POST /api/v1/users - 设置用户',
            'GET /api/v1/users/:id - 获取用户',
            'POST /api/v1/lists - 添加列表项',
            'GET /api/v1/lists/:list_key - 获取列表',
            'DELETE /api/v1/lists/:list_key/pop - 弹出列表项',
            'POST /api/v1/sets - 添加集合成员',
            'GET /api/v1/sets/:set_key - 获取集合成员',
            'GET /api/v1/redis/status - 获取 Redis 状态',
            'GET /api/v1/health - 健康检查'
        ]
    })

if __name__ == '__main__':
    print("Redis 哨兵模式 Web 应用启动中...")
    print("API 文档:")
    print("  POST   /api/v1/kv              - 设置键值对")
    print("  GET    /api/v1/kv/:key         - 获取键值对")
    print("  DELETE /api/v1/kv/:key         - 删除键")
    print("  GET    /api/v1/kv/:key/exists  - 检查键是否存在")
    print("  POST   /api/v1/users           - 设置用户")
    print("  GET    /api/v1/users/:id       - 获取用户")
    print("  POST   /api/v1/lists           - 添加列表项")
    print("  GET    /api/v1/lists/:list_key - 获取列表")
    print("  DELETE /api/v1/lists/:list_key/pop - 弹出列表项")
    print("  POST   /api/v1/sets            - 添加集合成员")
    print("  GET    /api/v1/sets/:set_key   - 获取集合成员")
    print("  GET    /api/v1/redis/status    - 获取 Redis 状态")
    print("  GET    /api/v1/health          - 健康检查")
    
    app.run(host='0.0.0.0', port=8080, debug=True)

File: test/go/web-app.go
package main

import (
	"context"
	"log"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
)

// WebApp Web 应用结构体
type WebApp struct {
	redis *RedisSentinelConfig
}

// NewWebApp 创建新的 Web 应用
func NewWebApp() *WebApp {
	return &WebApp{
		redis: RedisConfig,
	}
}

// Response 统一响应结构
type Response struct {
	Code    int         `json:"code"`
	Message string      `json:"message"`
	Data    interface{} `json:"data,omitempty"`
}

// User 用户结构体
type User struct {
	ID   int    `json:"id"`
	Name string `json:"name"`
	Age  int    `json:"age"`
}

// 成功响应
func success(c *gin.Context, data interface{}) {
	c.JSON(http.StatusOK, Response{
		Code:    200,
		Message: "success",
		Data:    data,
	})
}

// 错误响应
func errorResponse(c *gin.Context, code int, message string) {
	c.JSON(code, Response{
		Code:    code,
		Message: message,
	})
}

// 设置键值对
func (app *WebApp) setKeyValue(c *gin.Context) {
	var req struct {
		Key   string `json:"key" binding:"required"`
		Value string `json:"value" binding:"required"`
		TTL   int    `json:"ttl"` // 过期时间（秒）
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		errorResponse(c, http.StatusBadRequest, "参数错误: "+err.Error())
		return
	}

	ctx := context.Background()
	var expiration time.Duration
	if req.TTL > 0 {
		expiration = time.Duration(req.TTL) * time.Second
	}

	err := app.redis.Set(ctx, req.Key, req.Value, expiration)
	if err != nil {
		errorResponse(c, http.StatusInternalServerError, "设置失败: "+err.Error())
		return
	}

	success(c, gin.H{
		"key":   req.Key,
		"value": req.Value,
		"ttl":   req.TTL,
	})
}

// 获取键值对
func (app *WebApp) getKeyValue(c *gin.Context) {
	key := c.Param("key")
	if key == "" {
		errorResponse(c, http.StatusBadRequest, "键名不能为空")
		return
	}

	ctx := context.Background()
	value, err := app.redis.Get(ctx, key)
	if err != nil {
		errorResponse(c, http.StatusNotFound, "键不存在或获取失败: "+err.Error())
		return
	}

	success(c, gin.H{
		"key":   key,
		"value": value,
	})
}

// 删除键
func (app *WebApp) deleteKey(c *gin.Context) {
	key := c.Param("key")
	if key == "" {
		errorResponse(c, http.StatusBadRequest, "键名不能为空")
		return
	}

	ctx := context.Background()
	count, err := app.redis.Del(ctx, key)
	if err != nil {
		errorResponse(c, http.StatusInternalServerError, "删除失败: "+err.Error())
		return
	}

	success(c, gin.H{
		"key":   key,
		"count": count,
	})
}

// 检查键是否存在
func (app *WebApp) existsKey(c *gin.Context) {
	key := c.Param("key")
	if key == "" {
		errorResponse(c, http.StatusBadRequest, "键名不能为空")
		return
	}

	ctx := context.Background()
	count, err := app.redis.Exists(ctx, key)
	if err != nil {
		errorResponse(c, http.StatusInternalServerError, "检查失败: "+err.Error())
		return
	}

	success(c, gin.H{
		"key":    key,
		"exists": count > 0,
	})
}

// 设置用户信息
func (app *WebApp) setUser(c *gin.Context) {
	var user User
	if err := c.ShouldBindJSON(&user); err != nil {
		errorResponse(c, http.StatusBadRequest, "参数错误: "+err.Error())
		return
	}

	ctx := context.Background()
	userKey := "user:" + strconv.Itoa(user.ID)

	// 使用哈希存储用户信息
	_, err := app.redis.HSet(ctx, userKey, "id", user.ID, "name", user.Name, "age", user.Age)
	if err != nil {
		errorResponse(c, http.StatusInternalServerError, "设置用户失败: "+err.Error())
		return
	}

	success(c, gin.H{
		"user": user,
	})
}

// 获取用户信息
func (app *WebApp) getUser(c *gin.Context) {
	userIDStr := c.Param("id")
	userID, err := strconv.Atoi(userIDStr)
	if err != nil {
		errorResponse(c, http.StatusBadRequest, "用户ID格式错误")
		return
	}

	ctx := context.Background()
	userKey := "user:" + strconv.Itoa(userID)

	// 获取用户信息
	userData, err := app.redis.HGetAll(ctx, userKey)
	if err != nil {
		errorResponse(c, http.StatusInternalServerError, "获取用户失败: "+err.Error())
		return
	}

	if len(userData) == 0 {
		errorResponse(c, http.StatusNotFound, "用户不存在")
		return
	}

	// 解析用户数据
	age, _ := strconv.Atoi(userData["age"])
	user := User{
		ID:   userID,
		Name: userData["name"],
		Age:  age,
	}

	success(c, gin.H{
		"user": user,
	})
}

// 添加列表项
func (app *WebApp) addListItem(c *gin.Context) {
	var req struct {
		ListKey string   `json:"list_key" binding:"required"`
		Items   []string `json:"items" binding:"required"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		errorResponse(c, http.StatusBadRequest, "参数错误: "+err.Error())
		return
	}

	ctx := context.Background()

	// 转换为 interface{} 切片
	items := make([]interface{}, len(req.Items))
	for i, item := range req.Items {
		items[i] = item
	}

	count, err := app.redis.LPush(ctx, req.ListKey, items...)
	if err != nil {
		errorResponse(c, http.StatusInternalServerError, "添加列表项失败: "+err.Error())
		return
	}

	success(c, gin.H{
		"list_key": req.ListKey,
		"count":    count,
		"items":    req.Items,
	})
}

// 获取列表
func (app *WebApp) getList(c *gin.Context) {
	listKey := c.Param("list_key")
	if listKey == "" {
		errorResponse(c, http.StatusBadRequest, "列表键名不能为空")
		return
	}

	ctx := context.Background()
	items, err := app.redis.LRange(ctx, listKey, 0, -1)
	if err != nil {
		errorResponse(c, http.StatusInternalServerError, "获取列表失败: "+err.Error())
		return
	}

	success(c, gin.H{
		"list_key": listKey,
		"items":    items,
		"count":    len(items),
	})
}

// 弹出列表项
func (app *WebApp) popListItem(c *gin.Context) {
	listKey := c.Param("list_key")
	if listKey == "" {
		errorResponse(c, http.StatusBadRequest, "列表键名不能为空")
		return
	}

	ctx := context.Background()
	item, err := app.redis.RPop(ctx, listKey)
	if err != nil {
		errorResponse(c, http.StatusInternalServerError, "弹出列表项失败: "+err.Error())
		return
	}

	success(c, gin.H{
		"list_key": listKey,
		"item":     item,
	})
}

// 添加集合成员
func (app *WebApp) addSetMember(c *gin.Context) {
	var req struct {
		SetKey  string   `json:"set_key" binding:"required"`
		Members []string `json:"members" binding:"required"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		errorResponse(c, http.StatusBadRequest, "参数错误: "+err.Error())
		return
	}

	ctx := context.Background()

	// 转换为 interface{} 切片
	members := make([]interface{}, len(req.Members))
	for i, member := range req.Members {
		members[i] = member
	}

	count, err := app.redis.SAdd(ctx, req.SetKey, members...)
	if err != nil {
		errorResponse(c, http.StatusInternalServerError, "添加集合成员失败: "+err.Error())
		return
	}

	success(c, gin.H{
		"set_key": req.SetKey,
		"count":   count,
		"members": req.Members,
	})
}

// 获取集合成员
func (app *WebApp) getSetMembers(c *gin.Context) {
	setKey := c.Param("set_key")
	if setKey == "" {
		errorResponse(c, http.StatusBadRequest, "集合键名不能为空")
		return
	}

	ctx := context.Background()
	members, err := app.redis.SMembers(ctx, setKey)
	if err != nil {
		errorResponse(c, http.StatusInternalServerError, "获取集合成员失败: "+err.Error())
		return
	}

	success(c, gin.H{
		"set_key": setKey,
		"members": members,
		"count":   len(members),
	})
}

// 获取 Redis 状态
func (app *WebApp) getRedisStatus(c *gin.Context) {
	ctx := context.Background()

	// 获取主节点地址
	masterAddr, err := app.redis.GetMasterAddr(ctx)
	if err != nil {
		errorResponse(c, http.StatusInternalServerError, "获取主节点地址失败: "+err.Error())
		return
	}

	// 获取从节点地址
	slaveAddrs, err := app.redis.GetSlaveAddrs(ctx)
	if err != nil {
		errorResponse(c, http.StatusInternalServerError, "获取从节点地址失败: "+err.Error())
		return
	}

	// 获取哨兵信息
	sentinelInfo, err := app.redis.GetSentinelInfo(ctx)
	if err != nil {
		errorResponse(c, http.StatusInternalServerError, "获取哨兵信息失败: "+err.Error())
		return
	}

	success(c, gin.H{
		"master_addr":  masterAddr,
		"slave_addrs":  slaveAddrs,
		"sentinel_info": sentinelInfo,
	})
}

// 健康检查
func (app *WebApp) healthCheck(c *gin.Context) {
	ctx := context.Background()

	// 测试主节点连接
	_, err := app.redis.MasterClient.Ping(ctx).Result()
	if err != nil {
		errorResponse(c, http.StatusServiceUnavailable, "主节点连接失败: "+err.Error())
		return
	}

	// 测试从节点连接
	_, err = app.redis.SlaveClient.Ping(ctx).Result()
	if err != nil {
		errorResponse(c, http.StatusServiceUnavailable, "从节点连接失败: "+err.Error())
		return
	}

	success(c, gin.H{
		"status": "healthy",
		"timestamp": time.Now().Unix(),
	})
}

// 设置路由
func (app *WebApp) setupRoutes() *gin.Engine {
	r := gin.Default()

	// 中间件
	r.Use(gin.Logger())
	r.Use(gin.Recovery())

	// CORS 中间件
	r.Use(func(c *gin.Context) {
		c.Header("Access-Control-Allow-Origin", "*")
		c.Header("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")
		c.Header("Access-Control-Allow-Headers", "Content-Type, Authorization")

		if c.Request.Method == "OPTIONS" {
			c.AbortWithStatus(204)
			return
		}

		c.Next()
	})

	// API 路由组
	api := r.Group("/api/v1")
	{
		// 基本键值操作
		api.POST("/kv", app.setKeyValue)           // 设置键值对
		api.GET("/kv/:key", app.getKeyValue)       // 获取键值对
		api.DELETE("/kv/:key", app.deleteKey)      // 删除键
		api.GET("/kv/:key/exists", app.existsKey)  // 检查键是否存在

		// 用户操作
		api.POST("/users", app.setUser)     // 设置用户
		api.GET("/users/:id", app.getUser)  // 获取用户

		// 列表操作
		api.POST("/lists", app.addListItem)        // 添加列表项
		api.GET("/lists/:list_key", app.getList)   // 获取列表
		api.DELETE("/lists/:list_key/pop", app.popListItem) // 弹出列表项

		// 集合操作
		api.POST("/sets", app.addSetMember)        // 添加集合成员
		api.GET("/sets/:set_key", app.getSetMembers) // 获取集合成员

		// 系统操作
		api.GET("/redis/status", app.getRedisStatus) // 获取 Redis 状态
		api.GET("/health", app.healthCheck)         // 健康检查
	}

	return r
}

// 启动 Web 应用
func (app *WebApp) Run(port string) {
	r := app.setupRoutes()

	log.Printf("Web 应用启动在端口 %s", port)
	log.Printf("API 文档:")
	log.Printf("  POST   /api/v1/kv              - 设置键值对")
	log.Printf("  GET    /api/v1/kv/:key         - 获取键值对")
	log.Printf("  DELETE /api/v1/kv/:key         - 删除键")
	log.Printf("  GET    /api/v1/kv/:key/exists  - 检查键是否存在")
	log.Printf("  POST   /api/v1/users           - 设置用户")
	log.Printf("  GET    /api/v1/users/:id       - 获取用户")
	log.Printf("  POST   /api/v1/lists           - 添加列表项")
	log.Printf("  GET    /api/v1/lists/:list_key - 获取列表")
	log.Printf("  DELETE /api/v1/lists/:list_key/pop - 弹出列表项")
	log.Printf("  POST   /api/v1/sets            - 添加集合成员")
	log.Printf("  GET    /api/v1/sets/:set_key   - 获取集合成员")
	log.Printf("  GET    /api/v1/redis/status    - 获取 Redis 状态")
	log.Printf("  GET    /api/v1/health          - 健康检查")

	if err := r.Run(":" + port); err != nil {
		log.Fatal("启动 Web 应用失败:", err)
	}
}

func main() {
	// 创建 Web 应用
	app := NewWebApp()

	// 启动应用
	app.Run("8080")
}
